Lib/ufoLib/filesystem.py

```diff
@@ -0,0 +1,463 @@
+
+		mode and structure are both ignored if a
+		fs file system object is given for path.
+		"""
+		self._root = None
+		self._path = "<data stream>"
+		if isinstance(path, basestring):
+			self._path = path
+			if mode == "w":
+				if os.path.exists(path):
+					existingStructure = sniffFileStructure(path)
+					if structure is not None:
+						if structure != existingStructure:
+							raise UFOLibError("A UFO with a different structure already exists at the given path.")
+					else:
+						structure = existingStructure
+			elif mode == "r":
+				if not os.path.exists(path):
+					raise UFOLibError("The specified UFO doesn't exist.")
+				structure = sniffFileStructure(path)
+			if structure == "package":
+				path = OSFS(path)
+			elif structure == "zip":
+				if not haveFS:
+					raise UFOLibError("The fs module is required for reading and writing UFO ZIP.")
+				path = ZipFS(path, mode=mode, allow_zip_64=True, encoding="utf8")
+				roots = path.listdir("")
+				if not roots:
+					self._root = "contents"
+				elif len(roots) > 1:
+					raise UFOLibError("The UFO contains more than one root.")
+				else:
+					self._root = roots[0]
+		self._fs = path
+
+	def close(self):
+		self._fsClose()
+
+	# --------
+	# fs Calls
+	# --------
+
+	"""
+	All actual low-level file system interaction
+	MUST be done through these methods.
+
+	This is necessary because ZIPs will have a
+	top level root directory that packages will
+	not have.
+	"""
+
+	def _fsRootPath(self, path):
+		if self._root is None:
+			return path
+		return self.joinPath(self._root, path)
+
+	def _fsUnRootPath(self, path):
+		if self._root is None:
+			return path
+		return self.relativePath(path, self._root)
+
+	def _fsClose(self):
+		self._fs.close()
+
+	def _fsOpen(self, path, mode="r", encoding=None):
+		path = self._fsRootPath(path)
+		f = self._fs.open(path, mode, encoding=encoding)
+		return f
+
+	def _fsRemove(self, path):
+		path = self._fsRootPath(path)
+		self._fs.remove(path)
+
+	def _fsMakeDirectory(self, path):
+		path = self._fsRootPath(path)
+		self._fs.makedir(path)
+
+	def _fsRemoveDirectory(self, path):
+		path = self._fsRootPath(path)
+		self._fs.removedir(path)
+
+	def _fsMove(self, path1, path2):
+		if self.isDirectory(path1):
+			meth = self._fs.movedir
+		else:
+			meth = self._fs.move
+		path1 = self._fsRootPath(path1)
+		path2 = self._fsRootPath(path2)
+		meth(path1, path2)
+
+	def _fsExists(self, path):
+		path = self._fsRootPath(path)
+		return self._fs.exists(path)
+
+	def _fsIsDirectory(self, path):
+		path = self._fsRootPath(path)
+		return self._fs.isdir(path)
+
+	def _fsListDirectory(self, path):
+		path = self._fsRootPath(path)
+		return self._fs.listdir(path)
+
+	def _fsGetFileModificationTime(self, path):
+		path = self._fsRootPath(path)
+		info = self._fs.getinfo(path)
+		return info["modified_time"]
+
+	# -----------------
+	# Path Manipulation
+	# -----------------
+
+	def joinPath(self, *parts):
+		return os.path.join(*parts)
```

Comment:
Thanks. Both of these modules seem to offer things like pth.listdir() pth.chmod(). That seems to go against what we try to achieve by using fs, no? fs.path is probably the way to go.